File: c/uart.c
#include <stdio.h>
#include <gpiod.h>
#include "uart.h"


// GPIO Stuff -----------------------------------------------------------------------------

#ifndef	CONSUMER
#define	CONSUMER    "UART_DRIVER"
#endif

static struct gpiod_chip *chip8;
static struct gpiod_chip *chip7;
static struct gpiod_chip *chip5;
static struct gpiod_chip *chip1;
static struct gpiod_line *UART_CTRL_1_485_232_1;
static struct gpiod_line *UART_CTRL_1_485_232_2;
static struct gpiod_line *UART_CTRL_1_LB;
static struct gpiod_line *UART_CTRL_1_RXEN1;
static struct gpiod_line *UART_CTRL_1_DXEN1;
static struct gpiod_line *UART_CTRL_1_DXEN2;
static struct gpiod_line *UART_CTRL_1_RXEN2;
static struct gpiod_line *UART_CTRL_2_485_232_1;
static struct gpiod_line *UART_CTRL_2_485_232_2;
static struct gpiod_line *UART_CTRL_2_LB;
static struct gpiod_line *UART_CTRL_2_RXEN1;
static struct gpiod_line *UART_CTRL_2_DXEN1;
static struct gpiod_line *UART_CTRL_2_DXEN2;
static struct gpiod_line *UART_CTRL_2_RXEN2;

// control pins digital level
const int uart_modes[][4] = {
    [UART_1X_4X_6X_8X]             = {1, 1, 1, 1},
    [UART_1X_4X_6X_8Y_9Y]          = {1, 1, 1, 0},
    [UART_1X_4X_6Y_7Y_8X]          = {1, 1, 0, 1},
    [UART_1X_4X_6Y_7Y_8Y_9Y]       = {1, 1, 0, 0},
    [UART_1X_4Y_5Y_6X_8X]          = {1, 0, 1, 1},
    [UART_1X_4Y_5Y_6X_8Y_9Y]       = {1, 0, 1, 0},
    [UART_1X_4Y_5Y_6Y_7Y_8X]       = {1, 0, 0, 1},
    [UART_1X_4Y_5Y_6Y_7Y_8Y_9Y]    = {1, 0, 0, 0},
    [UART_1Y_2Y_4X_6X_8X]          = {0, 1, 1, 1},
    [UART_1Y_2Y_4X_6X_8Y_9Y]       = {0, 1, 1, 0},
    [UART_1Y_2Y_4X_6Y_7Y_8X]       = {0, 1, 0, 1},
    [UART_1Y_2Y_4X_6Y_7Y_8Y_9Y]    = {0, 1, 0, 0},
    [UART_1Y_2Y_4Y_5Y_6X_8X]       = {0, 0, 1, 1},
    [UART_1Y_2Y_4Y_5Y_6X_8Y_9Y]    = {0, 0, 1, 0},
    [UART_1Y_2Y_4Y_5Y_6Y_7Y_8X]    = {0, 0, 0, 1},
    [UART_1Y_2Y_4Y_5Y_6Y_7Y_8Y_9Y] = {0, 0, 0, 0},
};

static bool init_done = false;

#define CHECK_AND_GOTO_ON_FAILURE(expr, label) \
    do { \
        if (!(expr)) { \
            goto label; \
        } \
    } while (0)


struct gpiod_line* REQ_GPIO(struct gpiod_chip* chip, int line, int value) {

    struct gpiod_line* gpio = gpiod_chip_get_line(chip, line);
    if (!gpio) {
        printf("Error getting GPIO: %s line: %d\n", gpiod_chip_name(chip), line);
        return NULL;
    }
    
    int request_result = gpiod_line_request_output(gpio, "UART_DRVIER", value);
    if (request_result < 0) {
        printf("Error requesting GPIO: %s line: %d as output\n", gpiod_chip_name(chip), line);
        gpiod_line_release(gpio);
        return NULL;
    }
    
    int set_value_result = gpiod_line_set_value(gpio, value);
    if (set_value_result < 0) {
        printf("Error setting GPIO: %s line: %d value: %d\n", gpiod_chip_name(chip), line, value);
        gpiod_line_release(gpio);
        return NULL;
    }
    
    return gpio;
}


int UART_Open(uint8_t mode)
{
    if (init_done) return 0;

    chip8 = gpiod_chip_open_by_name("gpiochip8");
    CHECK_AND_GOTO_ON_FAILURE(chip8, cleanup);

    chip7 = gpiod_chip_open_by_name("gpiochip7");
    CHECK_AND_GOTO_ON_FAILURE(chip7, cleanup_chip8);

    chip5 = gpiod_chip_open_by_name("gpiochip5");
    CHECK_AND_GOTO_ON_FAILURE(chip5, cleanup_chip7);

    chip1 = gpiod_chip_open_by_name("gpiochip1");
    CHECK_AND_GOTO_ON_FAILURE(chip1, cleanup_chip5);

    // CTRL 1 (LTC2872)------------------------------------------------------------

    UART_CTRL_1_485_232_1 = REQ_GPIO(chip5, 13, 1);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_1_485_232_1, cleanup_chip1);

    UART_CTRL_1_485_232_2 = REQ_GPIO(chip5, 14, 1);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_1_485_232_2, cleanup_UART_CTRL_1_485_232_1);

    UART_CTRL_1_LB = REQ_GPIO(chip5, 25, 0);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_1_LB, cleanup_UART_CTRL_1_485_232_2);

    UART_CTRL_1_RXEN1 = REQ_GPIO(chip5, 27, 0);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_1_RXEN1, cleanup_UART_CTRL_1_LB);

    UART_CTRL_1_DXEN1 = REQ_GPIO(chip5, 28, 1);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_1_DXEN1, cleanup_UART_CTRL_1_RXEN1);

    UART_CTRL_1_DXEN2 = REQ_GPIO(chip5, 29, 1);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_1_DXEN2, cleanup_UART_CTRL_1_DXEN1);

    UART_CTRL_1_RXEN2 = REQ_GPIO(chip5, 30, 0);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_1_RXEN2, cleanup_UART_CTRL_1_DXEN2);

    // CTRL 2 (LTC2872)------------------------------------------------------------

    UART_CTRL_2_485_232_1 = REQ_GPIO(chip5, 31, 1);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_2_485_232_1, cleanup_UART_CTRL_1_RXEN2);

    UART_CTRL_2_485_232_2 = REQ_GPIO(chip1, 22, 1);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_2_485_232_2, cleanup_UART_CTRL_2_485_232_1);

    UART_CTRL_2_LB = REQ_GPIO(chip7, 29, 0);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_2_LB, cleanup_UART_CTRL_2_485_232_2);

    UART_CTRL_2_RXEN1 = REQ_GPIO(chip7, 31, 0);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_2_RXEN1, cleanup_UART_CTRL_2_LB);

    UART_CTRL_2_DXEN1 = REQ_GPIO(chip1, 0, 1);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_2_DXEN1, cleanup_UART_CTRL_2_RXEN1);

    UART_CTRL_2_DXEN2 = REQ_GPIO(chip1, 1, 1);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_2_DXEN2, cleanup_UART_CTRL_2_DXEN1);

    UART_CTRL_2_RXEN2 = REQ_GPIO(chip1, 2, 0);
    CHECK_AND_GOTO_ON_FAILURE(UART_CTRL_2_RXEN2, cleanup_UART_CTRL_2_DXEN2);

    init_done = true;

    UART_Change_Mode(mode);

    return 0;


cleanup_UART_CTRL_2_DXEN2:
    gpiod_line_release(UART_CTRL_2_DXEN2);
cleanup_UART_CTRL_2_DXEN1:
    gpiod_line_release(UART_CTRL_2_DXEN1);
cleanup_UART_CTRL_2_RXEN1:
    gpiod_line_release(UART_CTRL_2_RXEN1);
cleanup_UART_CTRL_2_LB:
    gpiod_line_release(UART_CTRL_2_LB);
cleanup_UART_CTRL_2_485_232_2:
    gpiod_line_release(UART_CTRL_2_485_232_2);
cleanup_UART_CTRL_2_485_232_1:
    gpiod_line_release(UART_CTRL_2_485_232_1);
cleanup_UART_CTRL_1_RXEN2:
    gpiod_line_release(UART_CTRL_1_RXEN2);
cleanup_UART_CTRL_1_DXEN2:
    gpiod_line_release(UART_CTRL_1_DXEN2);
cleanup_UART_CTRL_1_DXEN1:
    gpiod_line_release(UART_CTRL_1_DXEN1);
cleanup_UART_CTRL_1_RXEN1:
    gpiod_line_release(UART_CTRL_1_RXEN1);
cleanup_UART_CTRL_1_LB:
    gpiod_line_release(UART_CTRL_1_LB);
cleanup_UART_CTRL_1_485_232_2:
    gpiod_line_release(UART_CTRL_1_485_232_2);
cleanup_UART_CTRL_1_485_232_1:
    gpiod_line_release(UART_CTRL_1_485_232_1);
cleanup_chip1:
    gpiod_chip_close(chip1);
cleanup_chip5:
    gpiod_chip_close(chip5);
cleanup_chip7:
    gpiod_chip_close(chip7);
cleanup_chip8:
    gpiod_chip_close(chip8);
cleanup:
    init_done = false;
    return -1;
}


void UART_Close(void)
{
    if (init_done)
    {
        gpiod_line_release(UART_CTRL_2_RXEN2);
        gpiod_line_release(UART_CTRL_2_DXEN2);
        gpiod_line_release(UART_CTRL_2_DXEN1);
        gpiod_line_release(UART_CTRL_2_RXEN1);
        gpiod_line_release(UART_CTRL_2_LB);
        gpiod_line_release(UART_CTRL_2_485_232_2);
        gpiod_line_release(UART_CTRL_2_485_232_1);
        gpiod_line_release(UART_CTRL_1_RXEN2);  
        gpiod_line_release(UART_CTRL_1_DXEN2);
        gpiod_line_release(UART_CTRL_1_DXEN1);
        gpiod_line_release(UART_CTRL_1_RXEN1);
        gpiod_line_release(UART_CTRL_1_LB);
        gpiod_line_release(UART_CTRL_1_485_232_2);
        gpiod_line_release(UART_CTRL_1_485_232_1);
        gpiod_chip_close(chip1);
        gpiod_chip_close(chip5);
        gpiod_chip_close(chip7);
        gpiod_chip_close(chip8);

        init_done = false;
    }
}


void UART_Change_Mode(uint8_t mode)
{
    if (!init_done)
    {
        printf("UART_Open() function must be called first.\n");
        return;
    }

    if (mode < 1 || mode > 16)
    {
        printf("Invalid mode selected: %d\n", mode);
        return;
    }

    const int* config = uart_modes[mode];

    gpiod_line_set_value(UART_CTRL_1_485_232_1, config[0])
    gpiod_line_set_value(UART_CTRL_1_485_232_2, config[1])
    gpiod_line_set_value(UART_CTRL_2_485_232_1, config[2])
    gpiod_line_set_value(UART_CTRL_2_485_232_2, config[3])
    
}

File: python/uart.py
#!/usr/bin/python

import serial
import time
import os

from utils import *


def REQ_GPIO(chip, line, value):
    gpio = chip.get_line(line)
    gpio.request(consumer='UART_DRVIER', type=gpiod.LINE_REQ_DIR_OUT)
    gpio.set_value(value)
    return gpio


if not DEBUG:
    import gpiod

    chip8 = gpiod.Chip('gpiochip8')
    chip7 = gpiod.Chip('gpiochip7')
    chip5 = gpiod.Chip('gpiochip5')
    chip1 = gpiod.Chip('gpiochip1')

    # CTRL 1 (LTC2872)-------------------------------------------------------------------

    # 1 ---> uart1 (485/422), 0 ---> uart_1_2 (232)
    UART_CTRL_1_485_232_1 = REQ_GPIO(chip=chip5, line=13, value=1)

    # 1 ---> uart4 (485/422), 0 ---> uart_4_5 (232)
    UART_CTRL_1_485_232_2 = REQ_GPIO(chip=chip5, line=14, value=1)

    # Loop Back
    UART_CTRL_1_LB = REQ_GPIO(chip=chip5, line=25, value=0)

    # 0 ---> Enable RX uart1,2 
    UART_CTRL_1_RXEN1 = REQ_GPIO(chip=chip5, line=27, value=0)

    # 1 ---> Enable Tx uart1,2 
    UART_CTRL_1_DXEN1 = REQ_GPIO(chip=chip5, line=28, value=1)

    # 1 ---> Enable Tx uart4,5 
    UART_CTRL_1_DXEN2 = REQ_GPIO(chip=chip5, line=29, value=1)

    # 0 ---> Enable Rx uart4,5
    UART_CTRL_1_RXEN2 = REQ_GPIO(chip=chip5, line=30, value=0)

    # CTRL 2 (LTC2872)-------------------------------------------------------------------

    # 1 ---> uart6 (485/422), 0 ---> uart_6_7 (232)
    UART_CTRL_2_485_232_1 = REQ_GPIO(chip=chip5, line=31, value=1)

    # 1 ---> uart8 (485/422), 0 ---> uart_8_9 (232)
    UART_CTRL_2_485_232_2 = REQ_GPIO(chip=chip1, line=22, value=1)

    # Loop Back
    UART_CTRL_2_LB = REQ_GPIO(chip=chip7, line=29, value=0)

    # 0 ---> Enable RX uart6,7
    UART_CTRL_2_RXEN1 = REQ_GPIO(chip=chip7, line=31, value=0)

    # 1 ---> Enable TX uart6,7
    UART_CTRL_2_DXEN1 = REQ_GPIO(chip=chip1, line=0, value=1)

    # 1 ---> Enable TX uart8,9
    UART_CTRL_2_DXEN2 = REQ_GPIO(chip=chip1, line=1, value=1)

    # 0 ---> Enable RX uart8,9
    UART_CTRL_2_RXEN2 = REQ_GPIO(chip=chip1, line=2, value=0)


MODE_485_422 = "(485/422)"
MODE_232 = "(232)"


UARTS = [
    {'name': 'UART1', 'mode': MODE_485_422, 'dev': '/dev/ttyS0', 'gpio': [1, None, None, None]},
    {'name': 'UART1', 'mode': MODE_232,     'dev': '/dev/ttyS0', 'gpio': [0, None, None, None]},
    {'name': 'UART2', 'mode': MODE_232,     'dev': '/dev/ttyS1', 'gpio': [0, None, None, None]},
    {'name': 'UART4', 'mode': MODE_485_422, 'dev': '/dev/ttyS3', 'gpio': [None, 1, None, None]},
    {'name': 'UART4', 'mode': MODE_232,     'dev': '/dev/ttyS3', 'gpio': [None, 0, None, None]},
    {'name': 'UART5', 'mode': MODE_232,     'dev': '/dev/ttyS4', 'gpio': [None, 0, None, None]},
    {'name': 'UART6', 'mode': MODE_485_422, 'dev': '/dev/ttyS5', 'gpio': [None, None, 1, None]},
    {'name': 'UART6', 'mode': MODE_232,     'dev': '/dev/ttyS5', 'gpio': [None, None, 0, None]},
    {'name': 'UART7', 'mode': MODE_232,     'dev': '/dev/ttyS6', 'gpio': [None, None, 0, None]},
    {'name': 'UART8', 'mode': MODE_485_422, 'dev': '/dev/ttyS7', 'gpio': [None, None, None, 1]},
    {'name': 'UART8', 'mode': MODE_232,     'dev': '/dev/ttyS7', 'gpio': [None, None, None, 0]},
    {'name': 'UART9', 'mode': MODE_232,     'dev': '/dev/ttyS8', 'gpio': [None, None, None, 0]},
]

if not DEBUG:
    CURRENT_UART = serial.Serial(UARTS[0]['dev'], baudrate=DEFAULT_BAUDRATE)
else:
    CURRENT_UART = serial.Serial()


def change_current_uart_gui(idx):
    global CURRENT_UART
    CURRENT_UART.close()

    if not DEBUG:
        CURRENT_UART = serial.Serial(UARTS[idx]['dev'], baudrate=DEFAULT_BAUDRATE)
    else:
        CURRENT_UART = serial.Serial()



def set_ctrl_pins(config):
    if not DEBUG:
        ctrl_pins = [
            UART_CTRL_1_485_232_1,
            UART_CTRL_1_485_232_2,
            UART_CTRL_2_485_232_1,
            UART_CTRL_2_485_232_2,
        ]
        idx, value = next(((i, elem) for i, elem in enumerate(config) if elem is not None))
        ctrl_pins[idx].set_value(value)


def write_gui(data):
    if not DEBUG:
        global CURRENT_UART
        CURRENT_UART.write(data)


def read_gui():
    if not DEBUG:
        global CURRENT_UART
        return CURRENT_UART.read_all()


def write_console(uart_idx, data):
    if not DEBUG:
        ser = serial.Serial(UARTS[int(uart_idx)]['dev'])
        ser.write(data)
        print(f"{RED('SENT:')} {format_packet(data)}")
        time.sleep(0.1)
        print(f"{RED('RECV:')} {format_packet(ser.read_all())}\n")
        ser.close()
    else:
        print(f"{RED('SENT:')} {format_packet(data)}\n")


        
def UART_Close():
    if not DEBUG:
        UART_CTRL_1_485_232_1.release()
        UART_CTRL_1_485_232_2.release()
        UART_CTRL_1_LB.release()
        UART_CTRL_1_RXEN1.release()
        UART_CTRL_1_DXEN1.release()
        UART_CTRL_1_DXEN2.release()
        UART_CTRL_1_RXEN2.release()
        UART_CTRL_2_485_232_1.release()
        UART_CTRL_2_485_232_2.release()
        UART_CTRL_2_LB.release()
        UART_CTRL_2_RXEN1.release()
        UART_CTRL_2_DXEN1.release()
        UART_CTRL_2_DXEN2.release()
        UART_CTRL_2_RXEN2.release()

        chip8.close()
        chip7.close()
        chip5.close()
        chip1.close()


def main():
    clear_screen()
    while True:
        print("Choose Your Port:\n")
        for i, uart in enumerate(UARTS):
            print(f"{RED(str(i)+')')} {uart['name']} {uart['mode']} {uart['dev']}")
        print(f"{RED('B)')} {BLUE('Back')}\n")
        option = input("Option: ")
        print()
        if option.upper() == 'B':
            clear_screen()
            return
        elif option in [str(i) for i in range(len(UARTS))]:            
            set_ctrl_pins(UARTS[int(option)]['gpio'])
            while True:
                data_size = input('Enter data size in bytes: ')
                print()
                if data_size.isdigit():
                    data = os.urandom(int(data_size))
                    write_console(int(option), data)
                    break
                else:
                    clear_screen()
        else:
            clear_screen()
        





